File: Res/Sorting/RadixSort/radixSort.dart
void radixSort(List<int> list) {
  // Find the maximum number to determine the numbers of digits
  int max = list.reduce((a, b) => a > b ? a : b);

  // Perform counting sort for every digit
  for (int exp = 1; max ~/ exp > 0; exp *= 10) {
    countingSortByDigit(list, exp);
  }
}

void countingSortByDigit(List<int> list, int exp) {
  int n = list.length;
  List<int> output = List<int>.filled(n, 0);
  List<int> count = List<int>.filled(10, 0);

  for (int i = 0; i < n; i++) {
    int digit = (list[i] ~/ exp) % 10;
    count[digit]++;
  }
  print("");
  for (int i = 1; i < 10; i++) {
    count[i] += count[i - 1];
  }
  for (int i = n - 1; i >= 0; i--) {
    int digit = (list[i] ~/ exp) % 10;
    output[count[digit] - 1] = list[i];
    count[digit]--;
    print(count);
  }
  print("");

  for (int i = 0; i < n; i++) {
    list[i] = output[i];
  }
}

void main() {
  List<int> numbers = [170, 45, 75, 90, 802, 24, 2, 66];
  radixSort(numbers);
  print(numbers);
}

File: Res/Sorting/RadixSort/radixSort.py
def radix_sort(arr):
 
    max_value = max(arr)

    exp = 1
    while max_value // exp > 0:
        counting_sort_by_digit(arr, exp)
        exp *= 10
def counting_sort_by_digit(arr, exp):
    n = len(arr)
    
    output = [0] * n
    count = [0] * 10
    
    for i in range(n):
        digit = (arr[i] // exp) % 10
        count[digit] += 1
        
    for i in range(1, 10):
        count[i] += count[i - 1]
        
    for i in range(n - 1, -1, -1):
        digit = (arr[i] // exp) % 10
        output[count[digit] - 1] = arr[i]
        count[digit] -= 1
        
    for i in range(n):
        arr[i] = output[i]
                
        
b = [25,41,97,56,21,45,20]
radix_sort(b)
print(b)